compiler/resolution/functionResolution.cpp

```diff
@@ -9599,6 +9599,7 @@ static void resolveMoveForRhsCallExpr(CallExpr* call, Type* rhsType) {
 
 static void moveSetConstFlagsAndCheck(CallExpr* call, CallExpr* rhs) {
   if (rhs->isPrimitive(PRIM_GET_MEMBER) ||
+      rhs->isPrimitive(PRIM_GET_MEMBER_VALUE) ||
```

Comment:
This function sets the LHS according to the RHS, right? (`call` is usually a `PRIM_MOVE` so `call->get(1)` would be the LHS).

In that case, I think that this conditional needs to be aware of whether or not `rhs` is a ref. That's based on this comment:

```
  // PRIM_GET_MEMBER(_VALUE): aggregate, field
  // if the field is a ref:
  //   GET_MEMBER is invalid AST
  //   GET_MEMBER_VALUE returns the reference
  // if the field is not a ref
  //   GET_MEMBER returns a reference to the field
  //   GET_MEMBER_VALUE returns the field value
```

In particular the case I am concerned about here is that the field is not a reference and it's using `PRIM_GET_MEMBER_VALUE`. In that event, the value is copied to the LHS, and the const-ness of it is unrelated to the const-ness of the field.

Reply:
(the fix might be as simple as changing this line to `(rhs->isPrimitive(PRIM_GET_MEMBER_VALUE) && rhs->get(2)->qualType().isRef()) ||` )

Reply:
Unfortunately that gets us right back to the basic functionality tests not working (and thus the argument getting modified in the body of the function).  Maybe the qualified type of the tuple's fields are not getting set up correctly?  The code certainly behaves like it got a reference:
```
    ref read_x0[1608172]:_ref(int(64))[1146578]
    (1608177 'move' read_x0[1608172](1608174 '.v' args[1602037] x0[1602097]))
    val elt_x0[1608179]:int(64)[10]
    unknown coerce_tmp[1608187]:int(64)[10] "dead at end of block" "coerce temp" "insert auto destroy" "temp"
    (1608191 'move' coerce_tmp[1608187](1608189 'deref' read_x0[1608172]))
```
Maybe the comment is out of date?

Reply:
Okay, doing the following allows the tests to behave like we want:

```
(rhs->isPrimitive(PRIM_GET_MEMBER_VALUE) && call->get(1)->qualType().isRef()) ||
```

Does that feel like it's doing what you're looking for?

Reply:
Hmm, in your example, we know `x0` is a reference, and `PRIM_GET_MEMBER_VALUE` is returning that reference. So the thing to do is figure out how to make sure the field is a reference.

Checking `call->get(1)` for ref-ness doesn't seem right to me.

Reply:
What if we limit it to checking if the rhsBase is a tuple?  I suspect the value access will be treated the same for all tuples and the type of the tuple's field doesn't seem to be a ref on its own

Reply:
> What if we limit it to checking if the rhsBase is a tuple? I suspect the value access will be treated the same for all tuples 

I don't think so, because we have referential and value tuples.


> and the type of the tuple's field doesn't seem to be a ref on its own

That doesn't seem right. Maybe we are just asking if it's a ref in the wrong way or it isn't a ref yet or something. See

```
ref read_x0[1608172]:_ref(int(64))[1146578]
(1608177 'move' read_x0[1608172](1608174 '.v' args[1602037] x0[1602097]))
```

The type of `read_x0` would have been inferred from the field & `.v` / `PRIM_GET_MEMBER_VALUE` would return a value if it wasn't a ref.

What happens if you try to use `call->qualType().isRef()` ? Anyway I suspect there's something wrong in my/our understanding, but I can't understand why the earlier attempts would tell it that `x0` is not a ref in this case.

Reply:
I think you mean `rhs->qualType().isRef()`?  That returns true at least

Reply:
Yes, I did mean that. In that case, we should be able to use that to filter the `PRIM_GET_MEMBER_VALUE` case to ones where the const-ness should propagate.